corehq/apps/sms/views.py

```diff
@@ -665,36 +665,110 @@ def global_backend_map(request):
+    data = {}
+    for doc in iter_docs(CommCareCase.get_db(), case_ids):
+        if domain_obj.custom_case_username:
+            name = doc.get(domain_obj.custom_case_username, _('unknown'))
+        else:
+            name = doc.get('name', _('unknown'))
+        data[doc['_id']] = [name]
+    return data
+
+
+def get_mobile_worker_contact_info(domain_obj, user_ids):
+    data = {}
+    for doc in iter_docs(CommCareUser.get_db(), user_ids):
+        user = CommCareUser.wrap(doc)
+        data[user.get_id] = user.raw_username
+    return data
+
+
+def get_contact_info(domain):
+    verified_number_ids = VerifiedNumber.by_domain(domain, ids_only=True)
+    domain_obj = Domain.get_by_name(domain, strict=True)
+    case_ids = []
+    mobile_worker_ids = []
+    data = []
+    for doc in iter_docs(VerifiedNumber.get_db(), verified_number_ids):
+        owner_id = doc['owner_id']
+        if doc['owner_doc_type'] == 'CommCareCase':
+            case_ids.append(owner_id)
+            data.append([
+                None,
+                'case',
+                doc['phone_number'],
+                owner_id,
+            ])
+        elif doc['owner_doc_type'] == 'CommCareUser':
+            mobile_worker_ids.append(owner_id)
+            data.append([
+                None,
+                'mobile_worker',
+                doc['phone_number'],
+                owner_id,
+            ])
+    contact_data = get_case_contact_info(domain_obj, case_ids)
+    contact_data.update(get_mobile_worker_contact_info(domain_obj, mobile_worker_ids))
+    for row in data:
+        contact_info = contact_data.get(row[3])
+        row[0] = contact_info[0] if contact_info else _('(unknown)')
+    return data
+
+
+def format_contact_data(domain, data):
+    for row in data:
+        contact_id = row[3]
+        if row[1] == 'case':
+            row[1] = _('Case')
+            row.append(reverse('case_details', args=[domain, contact_id]))
+        elif row[1] == 'mobile_worker':
+            row[1] = _('Mobile Worker')
+            row.append(reverse(EditCommCareUserView.urlname, args=[domain, contact_id]))
+        else:
+            row.append('#')
+        row.append(reverse('sms_chat', args=[domain, contact_id]))
+
+
+@require_permission(Permissions.edit_data)
+@requires_privilege_with_fallback(privileges.OUTBOUND_SMS)
+def chat_contact_list(request, domain):
+    sEcho = request.GET.get('sEcho')
+    iDisplayStart = int(request.GET.get('iDisplayStart'))
+    iDisplayLength = int(request.GET.get('iDisplayLength'))
+    sSearch = request.GET.get('sSearch', '').strip()
+
+    data = get_contact_info(domain)
+    total_records = len(data)
+
+    if sSearch:
+        regex = re.compile('^.*%s.*$' % sSearch)
+        data = filter(lambda row: regex.match(row[0]) or regex.match(row[2]), data)
+    filtered_records = len(data)
+
+    data.sort(key=lambda row: row[0])
+    data = data[iDisplayStart:iDisplayStart + iDisplayLength]
```

Comment:
@snopoke Yeah, I had thought about the right way to do this for some time and this is kind of the best we can do right now without moving things to postgres (longer term, I'd like to move all of the messaging stuff to postgres).  We don't want to send all the data to the browser because that's what it was doing before and once you get over a few thousand contacts it completely overwhelms the js pagination and makes the page unusable.  But you bring up a good point about the caching - I just updated it so that it caches the results for 30 minutes.